### PR TITLE
Add class based filtering for YOLO via ":" delimeter

### PR DESCRIPTION
**Pull request by Nulliik**
As per https://github.com/mcmonkeyprojects/SwarmUI/issues/557 I tried to write this implementation myself. I request some testing of this, seemed good from my side

### CODE REVIEW

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py [hunk 1]
```diff
@@ -10,14 +10,15 @@ def INPUT_TYPES(cls):
...
+                "class_filter": ("STRING", { "default": "", "multiline": False }),
             },
         }
 
     CATEGORY = "SwarmUI/masks"
     RETURN_TYPES = ("MASK",)
     FUNCTION = "seg"
 
-    def seg(self, image, model_name, index):
+    def seg(self, image, model_name, index, class_filter):
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:21]: add a `= None` and move the input type into `optional` instead of `required`. That way pre-existing workflows with the node don't break

**Comment by mcmonkey4eva** [comment]
This new input should be added to (A) the `.md` docs (B) the inline docs (`prompttools.js`) and (C) the on-page UI for segmentation

(If you don't want to do all of those I can figure out docs after pulling)

File: src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs [hunk 2]
```diff
@@ -1223,6 +1223,11 @@ JArray doMaskShrinkApply(WorkflowGenerator g, JArray imgIn)
                     if (part.DataText.StartsWith("yolo-"))
                     {
                         string fullname = part.DataText.After("yolo-");
+                        string[] modelParts = fullname.Split(new[] { ':' }, 2, StringSplitOptions.None);
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs:1226]: just `Split(':', 2)`

File: src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs [hunk 3]
```diff
@@ -1223,6 +1223,11 @@ JArray doMaskShrinkApply(WorkflowGenerator g, JArray imgIn)
...
+
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs:1227]: avoid stray blank lines like this in the C#

**Comment by Nulliik** [comment]
Fixed the reviewed stuff and resolved empty mask appearing after filtering. I don't have time to make the docks currently, probably will return to in in the evening

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py [hunk 4]
```diff
@@ -28,17 +31,39 @@ def seg(self, image, model_name, index):
         from ultralytics import YOLO
         model = YOLO(model_path)
         results = model(img)
+        boxes = results[0].boxes
+        class_ids = results[0].boxes.cls.cpu().numpy() if boxes is not None else []
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:35]: this isn't using the `boxes` localvar but should be

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py [hunk 5]
```diff
@@ -28,17 +31,39 @@ def seg(self, image, model_name, index):
         from ultralytics import YOLO
         model = YOLO(model_path)
         results = model(img)
+        boxes = results[0].boxes
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:34]: make sure to test on non-class-based models btw, prior code did not attempt to read `boxes` at all unless it was a box model, not sure what effect that will have

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py [hunk 6]
```diff
@@ -28,17 +31,39 @@ def seg(self, image, model_name, index):
...
+        class_ids = results[0].boxes.cls.cpu().numpy() if boxes is not None else []
+        if class_filter and class_filter.strip():
+            selected_classes = [int(cls.strip()) for cls in class_filter.split(",") if cls.strip().isdigit()]
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: This is using numeric ID inputs? Don't classes have text labels?

Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: YOLO models output them with numeric ID, text labels are usually taken from somewhere else

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py [hunk 7]
```diff
@@ -28,17 +31,39 @@ def seg(self, image, model_name, index):
...
                 return (torch.zeros(1, image.shape[1], image.shape[2]), )
+            else:
+                if selected_classes:
+                    boxes = [box for i, box in enumerate(boxes) if class_ids[i] in selected_classes]
             masks = torch.zeros((len(boxes), image.shape[1], image.shape[2]), dtype=torch.float32, device="cpu")
             for i, box in enumerate(boxes):
                 x1, y1, x2, y2 = box.xyxy[0].tolist()
                 masks[i, int(y1):int(y2), int(x1):int(x2)] = 1.0
         else:
+            if isinstance(masks, list):
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:62]: why would this happen

File: src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs [hunk 8]
```diff
@@ -1223,6 +1223,10 @@ JArray doMaskShrinkApply(WorkflowGenerator g, JArray imgIn)
                     if (part.DataText.StartsWith("yolo-"))
                     {
                         string fullname = part.DataText.After("yolo-");
+                        string[] modelParts = fullname.Split(new[] {':'}, 2, StringSplitOptions.None);
```
Comment (mcmonkey4eva) [inline src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs:1226]: you edited the whitespace on this line instead of fixing it

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 5 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:34]: I am not totally sure how to do, YOLO always has some class, if model is trained to detect something it will be class "0".

Comment (mcmonkey4eva) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:34]: That's not correct. Yolo is a collection of many different model sub-classes, notably _segmentation_ models are preferred for Swarm, and _BBox_ models have historically been used often.
Swarm links recommended seg models in docs here https://github.com/mcmonkeyprojects/SwarmUI/blob/master/docs/Features/Prompt%20Syntax.md#automatic-segmentation-and-refining
the seg models emit neither classes nor boxes, they emit segment masks

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 6 above)
Comment (mcmonkey4eva) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: You earlier linked https://github.com/keremberke/awesome-yolov8-models as a class model example, and the top entry https://huggingface.co/keremberke/yolov8n-shoe-classification shows natively supported label set of `['adidas', 'converse', 'nike']`, not numeric IDs, and has sample code showing that `postprocess_classify_output` yields a format like `{"cat": 0.4, "dog": 0.6}`, so it seems text labels should be usable natively?

File: src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs (hunk 8 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/WorkflowGeneratorSteps.cs:1226]: Apologize, got it, it's obsolette

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 7 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:62]: I think forgot to use torch.stack somewhere before and forgot to remove it after I fixed, can't make it happen

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 6 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: Discussed as discord, subject to implement, but initially I will release it with class as id only

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 5 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:34]: Discussed in discord, apparently they do emit bbox, checked the https://docs.ultralytics.com/reference/engine/results/

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 4 above)
Comment (Nulliik) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:35]: Yep, my bad

**Comment by Nulliik** [comment]
Added docks, removed unnececary list check. Not sure about 
> "(C) the on-page UI for segmentation"
I wasn't able to locate where it is

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 6 above)
Comment (mcmonkey4eva) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: per Discord discussion, please implement name inputs. Only the in-UI helpers for names thing was "deal with it later cause it's complicated". Accepting name inputs should be simple

File: docs/Features/Prompt Syntax.md [hunk 9]
```diff
@@ -123,6 +123,7 @@
             - You can do this all in one prompt to individual refine specific faces separately
             - Without this, if there are multiple people, it will do a bulk segmented refine on all faces combined
             - Note the index order is sorted from leftmost detection to right
+        - To filter specific classes, append `:class_ids` where `class_ids` is a comma-separated list of class IDs, e.g., `yolo-modelnamehere:0,1,2`
```
Comment (mcmonkey4eva) [inline docs/Features/Prompt Syntax.md:126]: Er, this doesn't work.
Segment has a dynamic comma separated list of numbers as its input already. It needs a separator mark of some form. Whether that's `model:0:1:2,0.8,1` or `model:0,1,2:0.8,1`

File: src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py (hunk 6 above)
Comment (NuLL1011) [reply src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmComfyExtra/SwarmYolo.py:37]: Okay, will do

**Comment by Nulliik** [comment]
Fixed typo in docks, implemented new separator
